File: backend/app/schemas/bas.py
from datetime import datetime, date
from decimal import Decimal
from typing import Optional, List
from uuid import UUID
from pydantic import BaseModel, Field

from app.models.bas_period import BasStatus


class BasPeriodCreate(BaseModel):
    """Schema for creating a new BAS period."""
    start_date: date = Field(..., description="Period start date")
    end_date: date = Field(..., description="Period end date")


class BasPeriodResponse(BaseModel):
    """Schema for BAS period response data."""
    id: UUID
    start_date: date
    end_date: date
    status: BasStatus
    gst_collected: Decimal
    gst_paid: Decimal
    net_gst: Decimal
    total_sales: Decimal
    total_purchases: Decimal
    created_at: datetime
    updated_at: Optional[datetime] = None
    finalised_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class BasSummary(BaseModel):
    """Schema for GST/BAS summary calculations."""
    gst_collected: Decimal = Field(..., description="Total GST collected on sales")
    gst_paid: Decimal = Field(..., description="Total GST paid on purchases")
    net_gst: Decimal = Field(..., description="Net GST position (collected - paid)")
    total_sales: Decimal = Field(..., description="Total sales amount")
    total_purchases: Decimal = Field(..., description="Total purchases amount")
    transaction_count: int = Field(..., description="Number of transactions in period")
    unresolved_count: int = Field(..., description="Number of transactions with missing/ambiguous GST treatment")


class BasTransactionItem(BaseModel):
    """Schema for transaction item in BAS drill-down view."""
    id: UUID
    date: date
    description: str
    amount: Decimal
    gst_amount: Optional[Decimal]
    gst_treatment: Optional[str]
    category: Optional[str]
    supplier_name: Optional[str] = None
    reference: Optional[str] = None

    class Config:
        from_attributes = True


class BasExportData(BaseModel):
    """Schema for BAS export data."""
    period_id: UUID
    period_start: date
    period_end: date
    status: str
    generated_at: datetime
    summary: BasSummary
    gst_collected: Decimal
    gst_paid: Decimal
    net_gst: Decimal
    total_sales: Decimal
    total_purchases: Decimal


class BasPeriodListResponse(BaseModel):
    """Schema for paginated BAS period list."""
    items: List[BasPeriodResponse]
    total: int
    page: int
    page_size: int


class BasFinalizeRequest(BaseModel):
    """Schema for finalizing a BAS period."""
    confirm: bool = Field(True, description="Confirm finalization")


class BasFinalizeResponse(BaseModel):
    """Schema for BAS period finalization response."""
    message: str
    period_id: UUID
    status: BasStatus
    finalised_at: datetime

File: frontend/src/types/bas.ts
// BAS Period Status
export type BasStatus = 'draft' | 'finalised';

// BAS Period interfaces
export interface BasPeriod {
  id: string;
  start_date: string;
  end_date: string;
  status: BasStatus;
  gst_collected: string;
  gst_paid: string;
  net_gst: string;
  total_sales: string;
  total_purchases: string;
  created_at: string;
  updated_at?: string;
  finalised_at?: string;
}

// BAS Summary
export interface BasSummary {
  gst_collected: string;
  gst_paid: string;
  net_gst: string;
  total_sales: string;
  total_purchases: string;
  transaction_count: number;
  unresolved_count: number;
}

// BAS Transaction Item (for drill-down view)
export interface BasTransaction {
  id: string;
  date: string;
  description: string;
  amount: string;
  gst_amount: string | null;
  gst_treatment: string | null;
  category: string | null;
  supplier_name: string | null;
  reference: string | null;
}

// Create/Update interfaces
export interface BasPeriodCreate {
  start_date: string;
  end_date: string;
}

// Paginated response
export interface PaginatedBasPeriods {
  items: BasPeriod[];
  total: number;
  page: number;
  page_size: number;
}

export interface PaginatedBasTransactions {
  items: BasTransaction[];
  total: number;
  page: number;
  page_size: number;
}

// Period detail with summary
export interface BasPeriodDetail extends BasPeriod {
  summary: BasSummary;
}

// Export data
export interface BasExportData {
  period_id: string;
  period_start: string;
  period_end: string;
  status: string;
  generated_at: string;
  summary: BasSummary;
  gst_collected: string;
  gst_paid: string;
  net_gst: string;
  total_sales: string;
  total_purchases: string;
}

// Finalize response
export interface BasFinalizeResponse {
  message: string;
  period_id: string;
  status: BasStatus;
  finalised_at: string;
}

// Display options for status
export const BAS_STATUS_OPTIONS: { value: BasStatus; label: string; color: string }[] = [
  { value: 'draft', label: 'Draft', color: 'bg-amber-100 text-amber-700 border-amber-200' },
  { value: 'finalised', label: 'Finalised', color: 'bg-emerald-100 text-emerald-700 border-emerald-200' },
];

// Helper functions for display
export function getBasStatusLabel(status: BasStatus): string {
  const option = BAS_STATUS_OPTIONS.find(opt => opt.value === status);
  return option?.label || status;
}

export function getBasStatusColor(status: BasStatus): string {
  const option = BAS_STATUS_OPTIONS.find(opt => opt.value === status);
  return option?.color || 'bg-slate-100 text-slate-600 border-slate-200';
}

// Format helpers
export function formatCurrency(amount: string | number | null): string {
  if (amount === null || amount === undefined) return '-';
  const num = typeof amount === 'string' ? parseFloat(amount) : amount;
  if (isNaN(num)) return '-';
  return new Intl.NumberFormat('en-AU', {
    style: 'currency',
    currency: 'AUD',
  }).format(num);
}

export function formatDate(dateString: string | null): string {
  if (!dateString) return '-';
  const date = new Date(dateString);
  if (isNaN(date.getTime())) return '-';
  return date.toLocaleDateString('en-AU', {
    day: '2-digit',
    month: '2-digit',
    year: 'numeric',
  });
}

export function formatPeriodRange(startDate: string, endDate: string): string {
  return `${formatDate(startDate)} - ${formatDate(endDate)}`;
}
